modal/_container_entrypoint.py

```diff
@@ -60,6 +61,50 @@ class ImportedFunction:
     function: _Function
 
 
+class DaemonizedThreadPool:
```

Comment:
This isn't exactly new code, I just moved it out of the old `run_function_sync` implementation into it's own context manager class

Reply:
Thanks for splitting it out. But on a reading I think it might be a bit confusing this way because the exception-handling interface between DaemonizedThreadPool and its caller is tricky.

Usually splitting code up into classes helps with modularity. But in this case, the KeyboardInterrupt / BaseException-handling comments don't quite make sense anymore when reading the class itself, you need to go back into the code where it is used. Compare to the previous implementation where the error handling is perhaps easier to follow / it's also a pretty simple few lines of code?

```python
if imp_fun.input_concurrency > 1:
    # We can't use `concurrent.futures.ThreadPoolExecutor` here because in Python 3.11+, this
    # class has no workaround that allows us to exit the Python interpreter process without
    # waiting for the worker threads to finish. We need this behavior on SIGINT.

    import queue
    import threading

    spawned_workers = 0
    inputs: queue.Queue[Any] = queue.Queue()
    finished = threading.Event()

    def worker_thread():
        while not finished.is_set():
            try:
                args = inputs.get(timeout=1)
            except queue.Empty:
                continue
            try:
                run_input(*args)
            except BaseException:
                # This should basically never happen, since only KeyboardInterrupt is the only error that can
                # bubble out of from handle_input_exception and those wouldn't be raised outside the main thread
                pass
            inputs.task_done()

    for input_id, function_call_id, args, kwargs in container_io_manager.run_inputs_outputs(
        imp_fun.input_concurrency
    ):
        if spawned_workers < imp_fun.input_concurrency:
            threading.Thread(target=worker_thread, daemon=True).start()
            spawned_workers += 1
        inputs.put((input_id, function_call_id, args, kwargs))

    finished.set()
    inputs.join()
```

Reply:
Good point about the usage-coupled exception handling.

I thought it was a bit too messy to keep the implementation inlined in the input loop now that there is additional branching in there (and the threadpool is only used for a subset of iterations), and I like that the threadpool object becomes a sort of sync equivalent of `task_context.create_task()` in the loop.

I think I'll just refactor the DaemonizedThreadPool to be more akin to a "SyncTaskCancellationContext", and get rid of the context specific errors in favor of more general ones. I also think we can just skip the `inputs.join()` call for all exceptions (similar to the old implementation).